src/core/segment_reader.rs

```diff
@@ -280,6 +286,117 @@ impl SegmentReader {
+                        // In this case we need to map the potential fast field to the field name
+                        // accepted by the query parser.
+                        let create_canonical =
+                            !field_entry.is_expand_dots_enabled() && json_path.contains('.');
+                        if create_canonical {
+                            // Without expand dots enabled dots need to be escaped.
+                            let escaped_json_path = json_path.replace('.', "\\.");
+                            let full_path = format!("{}.{}", field_name, escaped_json_path);
+                            let full_path_unescaped = format!("{}.{}", field_name, &json_path);
+                            map_to_canonical.insert(full_path_unescaped, full_path.to_string());
+                            full_path
+                        } else {
+                            // With expand dots enabled, we can use '.' instead of '\u{1}'.
+                            json_path_sep_to_dot(&mut json_path);
+                            format!("{}.{}", field_name, json_path)
+                        }
+                    };
+                    indexed_fields.extend(
+                        encoded_fields_in_index
+                            .into_iter()
+                            .map(|(name, typ)| (build_path(&field_name, name), typ)),
+                    );
+                }
+            }
+        }
+        let mut fast_fields: Vec<(String, Type)> = self
+            .fast_fields()
+            .columnar()
+            .iter_columns()?
+            .map(|(mut field_name, handle)| {
+                json_path_sep_to_dot(&mut field_name);
+                // map to canonical path, to avoid similar but different entries.
+                // Eventually we should just accept '.' seperated for all cases.
+                let field_name = map_to_canonical
+                    .get(&field_name)
+                    .unwrap_or(&field_name)
+                    .to_string();
+
+                (field_name, Type::from(handle.column_type()))
+            })
+            .collect();
+        // Since the type is encoded differently in the fast field and in the inverted index,
+        // the order of the fields is not guaranteed to be the same. Therefore, we sort the fields.
+        // If we are sure that the order is the same, we can remove this sort.
+        indexed_fields.sort_unstable();
+        fast_fields.sort_unstable();
+
+        // Maybe too slow for the high cardinality case
+        fn is_field_stored(field_name: &str, schema: &Schema) -> bool {
+            schema
+                .find_field(field_name)
+                .map(|(field, _path)| schema.get_field_entry(field).is_stored())
+                .unwrap_or(false)
+        }
+
+        let merged: Vec<FieldMetadata> = indexed_fields
+            .into_iter()
+            .merge_join_by(fast_fields, |i, j| i.cmp(j))
```

Comment:
nitpick: we usually use `left / right` in the code base, or `lhs, rhs`.
```suggestion
            .merge_join_by(fast_fields, |i, j| i.cmp(j))
```